File: MidtermLabWork3/java/MP18.java
import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Scanner;

/**
 * MP18 - Remove rows with empty fields.
 * Student: CELADEZ, JED CEDRIC G.
 */
public class MP18 {

    public static void main(String[] args) {
        Scanner scanner = new Scanner(System.in);
        System.out.print("Enter dataset file path: ");
        String filePath = scanner.nextLine().trim();

        List<String> headers = new ArrayList<>();
        List<List<String>> rows = new ArrayList<>();

        try {
            loadDataset(filePath, headers, rows);
            runMP18(rows);
        } catch (IOException e) {
            System.out.println("Error: Unable to read file. " + e.getMessage());
        } catch (IllegalArgumentException e) {
            System.out.println("Error: " + e.getMessage());
        } finally {
            scanner.close();
        }
    }

    private static void runMP18(List<List<String>> rows) {
        List<List<String>> cleanedRows = new ArrayList<>();
        int rowsRemoved = 0;

        for (List<String> row : rows) {
            if (hasEmptyField(row)) {
                rowsRemoved++;
            } else {
                cleanedRows.add(row);
            }
        }

        System.out.println("\n--- Dataset Cleanup Summary ---");
        System.out.println("Removed " + rowsRemoved + " rows containing empty fields.");
        System.out.println("Remaining valid records: " + cleanedRows.size());

        System.out.println("\n--- Sample Records (First 5) ---");
        if (cleanedRows.isEmpty()) {
            System.out.println("No rows remain after cleanup.");
            return;
        }

        int sampleCount = Math.min(5, cleanedRows.size());
        for (int i = 0; i < sampleCount; i++) {
            System.out.println(String.join(" | ", cleanedRows.get(i)));
        }
    }

    private static boolean hasEmptyField(List<String> row) {
        for (String value : row) {
            if (value == null || value.trim().isEmpty()) {
                return true;
            }
        }
        return false;
    }

    private static void loadDataset(String filePath, List<String> headers, List<List<String>> rows)
            throws IOException {
        try (BufferedReader reader = new BufferedReader(new FileReader(filePath))) {
            String line;
            boolean headerFound = false;

            while ((line = reader.readLine()) != null) {
                if (line.trim().isEmpty()) {
                    continue;
                }

                List<String> cells = parseCsvLine(line);

                if (!headerFound) {
                    boolean isHeader = false;
                    for (String cell : cells) {
                        if ("candidate".equalsIgnoreCase(cell.trim())) {
                            isHeader = true;
                            break;
                        }
                    }

                    if (isHeader) {
                        headers.addAll(cells);
                        headerFound = true;
                    }
                    continue;
                }

                rows.add(cells);
            }
        }

        if (headers.isEmpty()) {
            throw new IllegalArgumentException("Dataset header row not found.");
        }
    }

    private static List<String> parseCsvLine(String line) {
        List<String> cells = new ArrayList<>();
        StringBuilder current = new StringBuilder();
        boolean inQuotes = false;

        for (int i = 0; i < line.length(); i++) {
            char ch = line.charAt(i);

            if (ch == '"') {
                if (inQuotes && i + 1 < line.length() && line.charAt(i + 1) == '"') {
                    current.append('"');
                    i++;
                } else {
                    inQuotes = !inQuotes;
                }
            } else if (ch == ',' && !inQuotes) {
                cells.add(current.toString());
                current.setLength(0);
            } else {
                current.append(ch);
            }
        }

        cells.add(current.toString());
        return cells;
    }
}

File: MidtermLabWork3/javascript/MP18.js
const fs = require('fs');
const readline = require('readline');

/**
 * MP18 – Remove rows with empty fields.
 * Student: CELADEZ, JED CEDRIC G.
 */

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

rl.question('Enter dataset file path: ', (filePath) => {
    if (!fs.existsSync(filePath)) {
        console.error('Error: File not found.');
        rl.close();
        return;
    }

    try {
        const fileContent = fs.readFileSync(filePath, 'utf8');
        const lines = fileContent.split(/\r?\n/);
        
        let headerFound = false;
        let headerRow = null;
        let cleanedRows = [];
        let rowsRemoved = 0;

        for (let line of lines) {
            if (line.trim() === "" || (!headerFound && !line.includes("Candidate"))) {
                continue;
            }

            const cells = parseCsvLine(line);

            if (!headerFound) {
                headerRow = cells;
                headerFound = true;
                continue;
            }

            // Check for empty or whitespace-only fields
            const hasEmpty = cells.some(cell => cell.trim().length === 0);

            if (!hasEmpty) {
                cleanedRows.push(cells);
            } else {
                rowsRemoved++;
            }
        }

        console.log(`\n--- Dataset Cleanup Summary ---`);
        console.log(`Removed ${rowsRemoved} rows containing empty fields.`);
        console.log(`Remaining valid records: ${cleanedRows.length}`);
        console.log(`\n--- Sample Records (First 5) ---`);

        if (cleanedRows.length === 0) {
            console.log('No rows remain after cleanup.');
        } else {
            if (headerRow) {
                console.log(headerRow.join(' | '));
            }

            cleanedRows.slice(0, 5).forEach(row => {
                console.log(row.join(' | '));
            });
        }

    } catch (err) {
        console.error('Error processing file:', err.message);
    } finally {
        rl.close();
    }
});

function parseCsvLine(line) {
    const cells = [];
    let current = '';
    let inQuotes = false;

    for (let i = 0; i < line.length; i++) {
        const ch = line[i];

        if (ch === '"') {
            if (inQuotes && line[i + 1] === '"') {
                current += '"';
                i++;
            } else {
                inQuotes = !inQuotes;
            }
        } else if (ch === ',' && !inQuotes) {
            cells.push(current);
            current = '';
        } else {
            current += ch;
        }
    }

    cells.push(current);
    return cells;
}
